packages/core/src/shared/utilities/resourceCache.ts

```diff
@@ -0,0 +1,177 @@
+    ) {}
+
+    abstract resourceProvider(): Promise<V>
+
+    async getResource(): Promise<V> {
+        const cachedValue = await this.tryLoadResourceAndLock()
+        const resource = cachedValue?.resource
+
+        // If cache is still fresh, return cached result, otherwise pull latest from the service
+        if (cachedValue && resource && resource.result) {
+            const duration = now() - resource.timestamp
+            if (duration < this.expirationInMilli) {
+                logger.info(`cache hit, duration(%sms) is less than expiration(%sms)`, duration, this.expirationInMilli)
+                // release the lock
+                await this.releaseLock(resource, cachedValue)
+                return resource.result
+            } else {
+                logger.info(
+                    `cached value is stale, duration(%sms) is older than expiration(%sms), invoking service API to pull the latest response`,
+                    duration,
+                    this.expirationInMilli
+                )
+            }
+        }
+
+        /**
+         * Possible paths here
+         *  1. cache doesn't exist.
+         *  2. cache exists but expired.
+         *  3. lock is held by other process and the waiting time is greater than the specified waiting time
+         */
+        logger.info(`cache miss, invoking service API to pull the latest response`)
+        try {
+            // Make the real network call / FS read to pull the resource
+            const latest = await this.resourceProvider()
+
+            // Update resource cache and release the lock
+            const r: Resource<V> = {
+                locked: false,
+                timestamp: now(),
+                result: latest,
+            }
+            logger.info(`doen loading the latest of resource(%s), updating resource cache`, this.key)
+            await this.releaseLock(r)
+            return latest
+        } catch (e) {
+            logger.info(
+                `encountered unexpected error while loading the latest of resource(%s), releasing resource lock`,
+                this.key
+            )
+            await this.releaseLock()
+            throw e
+        }
+    }
+
+    // This method will lock the resource so other callers have to wait until the lock is released, otherwise will return undefined if it times out
+    private async tryLoadResourceAndLock(): Promise<GlobalStateSchema<V> | undefined> {
+        const _acquireLock = async () => {
+            const cachedValue = this.readCacheOrDefault()
+
+            if (!cachedValue.resource.locked) {
+                await this.lockResource(cachedValue)
+                return cachedValue
+            }
+
+            return undefined
+        }
+
+        const lock = await waitUntil(async () => {
+            const lock = await _acquireLock()
+            logger.info(`try obtain resource cache read lock %s`, lock)
```

Comment:
is stringified `lock` going to log anything useful?
```suggestion
            logger.info(`try obtain resource cache read lock %s`, this.key)
```

Reply:
yea that's a good callout, changed it to print this.key